### PR TITLE
Allow overriding of keycloak root credentials for `2024.11.1` upgrade path

### LINKED ISSUE #2833: [BUG] - Nebari Upgrade Keycloak Client Cannot Connect if initial_root_password changed
### Describe the bug

When upgrading to Nebari 2024.11.1rc1, there is a step where the upgrade command will try to create a new client role and assign groups in Keycloak.  It authenticates by accessing `security.keycloak.initial_root_password` from Nebari config

However, it is a best practice in a production environment to change that value in the config file after the initial deployment... otherwise a secret would be committed to CICD

The resulting error is:

```
Would you like Nebari to assign the corresponding role/scopes to all of your current groups automatically? [y/N] (N): y
...
ValueError: Failed to connect to Keycloak server: 401: b'{"error":"invalid_grant","error_description":"Invalid user credentials"}'
```

### Expected behavior

If you choose "Y" to assign corresponding roles scopes, the upgrade step should succeed even if your Keycloak root PW isnt' saved in plain text

### OS and architecture in which you are running Nebari

Mac Sequoia / Apple silicon

### How to Reproduce the problem?

Start with Nebari 2024.7.1 and `security.keycloak.initial_root_password` should no longer be valid (i.e. you have manually reset it from within the Keycloak web app)

When you install Nebari 2024.11.1rc1 and run `Nebari upgrade`, choose 'Y' at the prompt to create new client role and assign groups for shared directory access

Keycloak client fails to authenticate

### Command output

```bash session
Would you like Nebari to assign the corresponding role/scopes to all of your current groups automatically? [y/N] (N): y
...
ValueError: Failed to connect to Keycloak server: 401: b'{"error":"invalid_grant","error_description":"Invalid user credentials"}'
```

### Versions and dependencies used.

Nebari 2024.11.1rc1

### Compute environment

AWS

### Integrations

_No response_

### Anything else?

Probably, prompting the user to enter valid Keycloak credentials in the upgrade CLI is a good solution

In one of our implementations, we keep the valid Keycloak root password in AWS Secret Manager, and from there it could be accessed programmatically with valid AWS credentials, and then used to generate an oAuth token to talk to Keycloak's REST API.  This could be another option, but it would need cloud specific implementations and also a clear administrative workflow on when/where/how to update the Keycloak password secret (and be sure the AWS secret stays in sync with the actual password value).  So probably user prompting is easier...

Here's some snippets from a script which uses this approach:

```
NEBARI_CONFIG=${DIR}/../nebari-config.yaml

PROJECT_NAME=$(yq '.project_name' ${NEBARI_CONFIG})
REGION=$(yq '.amazon_web_services.region' ${NEBARI_CONFIG})

KC_SECRET_JSON=$(aws secretsmanager get-secret-value --secret-id ${PROJECT_NAME}/keycloak/admin --region ${REGION} | jq -rc '.SecretString')
KC_USERNAME=$(echo ${KC_SECRET_JSON} | jq -rc '.username')
KC_PASSWORD=$(echo ${KC_SECRET_JSON} | jq -rc '.password')
KC_URL=https://$(yq '.domain' ${NEBARI_CONFIG})/auth

if [ "${1}" == "" ]; then
  echo "ERROR: yaml file path required"
  exit 1
fi

IMPORT_YAML=$(realpath ${1})

echo "Importing users from ${IMPORT_YAML}..."

for username in $( yq '.[] | key' ${IMPORT_YAML} ); do
  first=$( yq ".\"${username}\".first" ${IMPORT_YAML} )
  last=$( yq ".\"${username}\".last" ${IMPORT_YAML} )
  groups=$( yq -I0 -M -o=json ".\"${username}\".groups" ${IMPORT_YAML} )

  ACCESS_TOKEN=$(curl -s -X POST -H 'Content-Type: application/x-www-form-urlencoded' --data "username=${KC_USERNAME}&password=${KC_PASSWORD}&grant_type=password&client_id=admin-cli" ${KC_URL}/realms/master/protocol/openid-connect/token | jq -rc '.access_token')
  GET_USER_RESP=$(curl -s -H "Content-Type: application/json" -H "Authorization: Bearer ${ACCESS_TOKEN}" ${KC_URL}/admin/realms/nebari/users?username=${username}&exact=true)
  ```

### PR DESCRIPTION
**Pull request by viniciusdc**
## Reference Issues or PRs
closes #2833 

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

### CODE REVIEW

File: src/_nebari/upgrade.py [hunk 1]
```diff
@@ -1297,10 +1298,16 @@ def _version_specific_upgrade(
 
             urllib3.disable_warnings()
 
+            keycloak_username = os.environ.get("KEYCLOAK_ADMIN_USERNAME", "root")
```
Comment (viniciusdc) [inline src/_nebari/upgrade.py:1301]: I am including a first layer check for the password, in case the user changed it

**Comment by viniciusdc** [comment]
needs to be validated if the keycloak_admin.get_server_info works as a first layer credential check

File: src/_nebari/upgrade.py [hunk 2]
```diff
@@ -1297,11 +1298,34 @@ def _version_specific_upgrade(
...
+                if "invalid_grant" in str(e):
+                    print(
+                        "[red bold]Failed to connect to the Keycloak server.[/red bold]\n"
+                        "[yellow]This may occur if the default admin credentials have been changed for security reasons.[/yellow]\n"
+                        "Please update the [bold]KEYCLOAK_ADMIN_USERNAME[/bold] and [bold]KEYCLOAK_ADMIN_PASSWORD[/bold] environment variables with the new credentials and try again."
+                    )
+                    exit()
+                else:
+                    # Handle other exceptions
+                    print(f"[red bold]An unexpected error occurred: {e}[/red bold]")
```
Comment (Adam-D-Lewis) [inline src/_nebari/upgrade.py:1326]: Will the whole stack trace be in red bold?

File: src/_nebari/upgrade.py [hunk 3]
```diff
@@ -1297,11 +1298,34 @@ def _version_specific_upgrade(
...
             )
+
+            try:
+                # Quick test to connect to Keycloak
+                keycloak_admin.get_server_info()
+            except ValueError as e:
+                if "invalid_grant" in str(e):
+                    print(
+                        "[red bold]Failed to connect to the Keycloak server.[/red bold]\n"
+                        "[yellow]This may occur if the default admin credentials have been changed for security reasons.[/yellow]\n"
```
Comment (Adam-D-Lewis) [inline src/_nebari/upgrade.py:1320]: I feel like this makes it sound like the user did something wrong or optional when in reality changing the security credentials should be the default behaviors. I'll suggest an edit.

File: src/_nebari/upgrade.py [hunk 4]
```diff
@@ -1297,11 +1298,34 @@ def _version_specific_upgrade(
...
+                        "Please update the [bold]KEYCLOAK_ADMIN_USERNAME[/bold] and [bold]KEYCLOAK_ADMIN_PASSWORD[/bold] environment variables with the new credentials and try again."
```
Comment (Adam-D-Lewis) [inline src/_nebari/upgrade.py:1321]:
```suggestion
                        "[red bold]Failed to connect to the Keycloak server.[/red bold]\n"
                        "Please set the [bold]KEYCLOAK_ADMIN_USERNAME[/bold] and [bold]KEYCLOAK_ADMIN_PASSWORD[/bold] environment variables with the root credentials and try again."
```

**Review by Adam-D-Lewis — COMMENTED** [review]
see comments

File: src/_nebari/upgrade.py (hunk 2 above)
Comment (Adam-D-Lewis) [inline src/_nebari/upgrade.py:1326]:
```suggestion
                    print(f"[red bold]An unexpected error occurred: {repr(e)}[/red bold]")
```

Comment (Adam-D-Lewis) [reply src/_nebari/upgrade.py:1326]: I see it doesn't print out the stack trace.  I agree with the comments on https://stackoverflow.com/a/1483488/9848141 that repr would be better than str(e) to give more info about what is happening.

**Comment by viniciusdc** [comment]
This is how it looks like now:
Captura de tela de 2024-11-12 13 40 57

**Review by Adam-D-Lewis — APPROVED** [review]

**Pull request merged**